[Snapshot download] allow some filenames to be ignored (#566)
huggingface/huggingface_hub: docs/hub/how-to-downstream.md

>>> snapshot_download(repo_id="lysandre/arxiv-nlp", revision="main")
```

> In general, it is usually better to manually download files with `hf_hub_download` (if you already know which files you need) to avoid redownloading an entire repository. `snapshot_download` is helpful when your library's downloading utility is a helper, and unaware of which files need to be downloaded.

Reviewer: ```suggestion
In general, it is usually better to manually download files with `hf_hub_download` (if you already know the file name) to avoid re-downloading an entire repository. `snapshot_download` is helpful when your library's downloading utility is a helper, and unaware of which files need to be downloaded.
```

> To only download some files of a repository with `snapshot_download`, you can make 

Reviewer: Help set some context for why this is useful :)

```suggestion
However, you don't want to always download the contents of an entire repository with `snapshot_download`. Even if you don't know the file name and only know the file type, you can download specific files with `allow_regex` and `ignore_regex`.
```

> make use of the `allow_regex` and `ignore_regex` arguments to specific 

Reviewer 2: ```suggestion
use of the `allow_regex` and `ignore_regex` arguments to specify 
```

which files shall be downloaded.

> Both a single regex or a list of regex are accepted for `allow_regex` and `ignore_regex`. 

Reviewer 2: ```suggestion
Both a single regex or a list of regexes are accepted for `allow_regex` and `ignore_regex`. 
```

> Both a single regex or a list of regex are accepted for `allow_regex` and `ignore_regex`. 

Reviewer: ```suggestion
`allow_regex` and `ignore_regex` accept either a single regex or a list of regexes. 
```

> The regex matching is based on [`fnmatch`](https://docs.python.org/3/library/fnmatch.html) which means that it provides support for Unix shell-style wildcards.

Reviewer: ```suggestion
The regex matching is based on [`fnmatch`](https://docs.python.org/3/library/fnmatch.html) which means it provides support for Unix shell-style wildcards.
```

> To download, *e.g.*, only the JSON configuration files, you could make use of `allow_regex`:

Reviewer: ```suggestion
For example, you can use `allow_regex` to only download JSON configuration files:
```

```python
>>> from huggingface_hub import snapshot_download
>>> snapshot_download(repo_id="lysandre/arxiv-nlp", allow_regex="*.json")
```

> If, on the other hand, you might want to always exclude any files of `.msgpack` 

Reviewer 2: ```suggestion
If, on the other hand, you want to always exclude any files of `.msgpack` 
```

> If, on the other hand, you might want to always exclude any files of `.msgpack` 

Reviewer: ```suggestion
On the other hand, `ignore_regex` can be used to exclude certain files from being downloaded. The following example ignores the `.msgpack` and `.h5` file extensions:
```

or `.h5` extensions, you could make use of `ignore_regex`:

```python
>>> from huggingface_hub import snapshot_download
>>> snapshot_download(repo_id="lysandre/arxiv-nlp", ignore_regex=["*.msgpack", "*.h5"])
```

Passing a regex can be especially useful when repositories contain files that 

> are are never expected to be downloaded by `snapshot_download`.

Reviewer: ```suggestion
are never expected to be downloaded by `snapshot_download`.
```
